server/new-commiter.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/google/go-github/v25/github"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type contributionInfo struct {
+	contributionWeek time.Time
+	repo             string
+}
+
+type firstContributionInfo struct {
+	author string
+	date   time.Time
+	commit string
+}
+
+func (p *Plugin) executeNewCommiterCommand(commandArgs []string, args *model.CommandArgs) *model.AppError {
+	if len(commandArgs) != 2 {
+		return &model.AppError{
+			Id:         "Need two arguments",
+			StatusCode: http.StatusBadRequest,
+			Where:      "p.ExecuteCommand",
+		}
+	}
+
+	organisation := commandArgs[0]
+
+	since, err := time.Parse(shortFormWithDay, commandArgs[1])
+	if err != nil {
+		return &model.AppError{
+			Id:         fmt.Sprintf("Failed to parse since time: %v", err.Error()),
+			StatusCode: http.StatusBadRequest,
+			Where:      "p.ExecuteCommand",
+		}
+	}
+
+	org, _, err := p.client.Organizations.Get(context.Background(), organisation)
+	if err != nil {
+		return &model.AppError{
+			Id:         "Failed to fetch data",
+			StatusCode: http.StatusBadRequest,
+			Where:      "p.ExecuteCommand",
+		}
+	}
+
+	attachments := []*model.SlackAttachment{{
+		Title:      "Fetching new commiters since " + since.Format(shortFormWithDay),
+		Text:       waitText,
+		AuthorName: organisation,
+		AuthorIcon: org.GetAvatarURL(),
+		AuthorLink: fmt.Sprintf("https://github.com/%v", organisation),
+	}}
+
+	loadingPost := &model.Post{
+		ChannelId: args.ChannelId,
+		UserId:    p.botUserID,
+	}
+	model.ParseSlackAttachment(loadingPost, attachments)
+
+	loadingPost, appErr := p.API.CreatePost(loadingPost)
+	if appErr != nil {
+		return appErr
+	}
+
+	go p.updateNewCommitersPost(loadingPost, args.UserId, organisation, since)
+
+	return nil
+}
+
+func (p *Plugin) updateNewCommitersPost(post *model.Post, userID, org string, since time.Time) {
+	contributors, err := p.fetchContributorsFromOrg(org)
+	if err != nil {
+		p.API.LogError("failed to fetch data", "err", err.Error())
+
+		var message string
+		if _, ok := err.(*github.RateLimitError); ok {
+			message = "Hit rate limit. Please try again later."
+		} else {
+			message = "Failed to fetch data:" + err.Error()
+		}
+		post.Props["attachments"].([]*model.SlackAttachment)[0].Text = message
+	} else {
+		firstContributions := p.findFirstContributions(contributors)
+		p.filterContributionsByWeek(firstContributions, since)
+
+		var result []*firstContributionInfo
+		for commiter, contribution := range firstContributions {
+			commits, _ := p.fetchCommitsFromRepoByAuthorAndWeek(org, contribution.repo, commiter, contribution.contributionWeek)
+			firstCommit := commits[0]
```

Comment:
This line crashes the plugin a couple of times when I tested:
```
DEBU[2020-06-10T12:16:09.3385925+02:00] panic: runtime error: index out of range [0] with length 0  caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.338687+02:00]                                               caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.3386934+02:00] goroutine 1743 [running]:                     caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.3386972+02:00] main.(*Plugin).updateNewCommitersPost(0xc00003a420, 0xc0038b1b00, 0xc0038f1e40, 0x1a, 0xc0027caa98, 0xa, 0x0, 0xed59dd800, 0x0)  caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.338702+02:00] 	/home/sumacheb/src/mattermost/plugins/community/server/new-commiter.go:102 +0x10da  caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.3387063+02:00] created by main.(*Plugin).executeNewCommiterCommand  caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.3387089+02:00] 	/home/sumacheb/src/mattermost/plugins/community/server/new-commiter.go:75 +0x5d8  caller="plugin/hclog_adapter.go:53" plugin_id=com.mattermost.community
DEBU[2020-06-10T12:16:09.3488672+02:00] plugin process exited                         caller="plugin/hclog_adapter.go:51" plugin_id=com.mattermost.community wrapped_extras="pathplugins/com.mattermost.community/server/dist/plugin-linux-amd64pid78386errorexit status 2"
```
I did run `/community new-commiter mattermost 2020-01-01`

Reply:
Thanks for investigating into this :+1: I'm a bit surprised how fast that cap is hit. E.g. `/community new-committer mattermost 2020-04-01` already fails for me. Maybe the list of contributors should be a separate post?

Reply:
As this cap is hit mainly by commit URLs (I guess), this error happens even when contributors list is a separate post. I tried updating existing post with just number of new commiters and creating new post for their contributions' details, `/community new-committer mattermost 2020-04-13` worked but `/community new-committer mattermost 2020-04-01` still fails

Reply:
"normal" post have a limit of 16000 characters IIRC. That might help here. Post attachments can only be 9000 characters long.

Reply:
It helped, now getting new committers is possible since 2020-03-02

Reply:
Not bad. Mind pushing the changes to I can test them myself?